src/System.IO.FileSystem/src/System/IO/FindEnumerable.Windows.cs

```diff
@@ -0,0 +1,252 @@
+        /// </summary>
+        public static IntPtr CreateDirectoryHandle(string path)
+        {
+            IntPtr handle = Interop.Kernel32.CreateFile_IntPtr(
+                path,
+                Interop.Kernel32.FileOperations.FILE_LIST_DIRECTORY,
+                FileShare.ReadWrite | FileShare.Delete,
+                FileMode.Open,
+                Interop.Kernel32.FileOperations.FILE_FLAG_BACKUP_SEMANTICS);
+
+            if (handle == IntPtr.Zero || handle == (IntPtr)(-1))
+            {
+                // Historically we throw directory not found rather than file not found
+                int error = Marshal.GetLastWin32Error();
+                if (error == Interop.Errors.ERROR_FILE_NOT_FOUND)
+                    error = Interop.Errors.ERROR_PATH_NOT_FOUND;
+
+                throw Win32Marshal.GetExceptionForWin32Error(error, path);
+            }
+
+            return handle;
+        }
+
+        public IEnumerator<TResult> GetEnumerator()
+        {
+            if (Interlocked.Exchange(ref _enumeratorCreated, 1) == 0)
+            {
+                return this;
+            }
+            else
+            {
+                return new FindEnumerable<TResult, TState>(_originalUserPath, _originalFullPath, _transform, _predicate, _state, _recursive);
+            }
+        }
+
+        private void Initialize()
+        {
+            _currentPath = _originalFullPath;
+            _buffer = ArrayPool<byte>.Shared.Rent(4096);
+            _pinnedBuffer = GCHandle.Alloc(_buffer, GCHandleType.Pinned);
+            if (_recursive)
+                _pending = new Queue<(IntPtr, string)>();
+            _directoryHandle = CreateDirectoryHandle(_originalFullPath);
+        }
+
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        public TResult Current => _current;
+
+        object IEnumerator.Current => Current;
+
+        public bool MoveNext()
+        {
+            if (_lastEntryFound)
+                return false;
+
+            lock (_lock)
+            {
+                if (_lastEntryFound)
+                    return false;
+
+                RawFindData findData = default;
+                do
+                {
+                    FindNextFile();
+                    if (!_lastEntryFound && _info != null)
+                    {
+                        // If needed, stash any subdirectories to process later
+                        if (_recursive && (_info->FileAttributes & FileAttributes.Directory) != 0
+                            && !PathHelpers.IsDotOrDotDot(_info->FileName))
+                        {
+                            string subDirectory = PathHelpers.CombineNoChecks(_currentPath, _info->FileName);
+                            IntPtr subDirectoryHandle = CreateDirectoryHandle(subDirectory);
+                            try
+                            {
+                                _pending.Enqueue((subDirectoryHandle, subDirectory));
```

Comment:
They try/catch is because Enqueue might allocate and thus might throw? Might be worth a comment.

Reply:
Yes, I'll add the comment